filament/backend/src/webgpu/WebGPUPipelineCreation.cpp

```diff
@@ -0,0 +1,254 @@
+            return wgpu::BlendFactor::Src;
+        case BlendFunction::ONE_MINUS_SRC_COLOR:
+            return wgpu::BlendFactor::OneMinusSrc;
+        case BlendFunction::DST_COLOR:
+            return wgpu::BlendFactor::Dst;
+        case BlendFunction::ONE_MINUS_DST_COLOR:
+            return wgpu::BlendFactor::OneMinusDst;
+        case BlendFunction::SRC_ALPHA:
+            return wgpu::BlendFactor::SrcAlpha;
+        case BlendFunction::ONE_MINUS_SRC_ALPHA:
+            return wgpu::BlendFactor::OneMinusSrcAlpha;
+        case BlendFunction::DST_ALPHA:
+            return wgpu::BlendFactor::DstAlpha;
+        case BlendFunction::ONE_MINUS_DST_ALPHA:
+            return wgpu::BlendFactor::OneMinusDstAlpha;
+        case BlendFunction::SRC_ALPHA_SATURATE:
+            return wgpu::BlendFactor::SrcAlphaSaturated;
+    }
+}
+
+}// namespace
+
+wgpu::RenderPipeline createWebGPURenderPipeline(wgpu::Device const& device,
+        WGPUProgram const& program, WGPUVertexBufferInfo const& vertexBufferInfo,
+        wgpu::PipelineLayout const& layout, RasterState const& rasterState,
+        StencilState const& stencilState, PolygonOffset const& polygonOffset,
+        PrimitiveType primitiveType, wgpu::TextureFormat colorFormat,
+        wgpu::TextureFormat depthFormat) {
+    assert_invariant(program.vertexShaderModule);
+    const wgpu::DepthStencilState depthStencilState {
+        .format = depthFormat,
+        .depthWriteEnabled = rasterState.depthWrite,
+        .depthCompare = toWebGPU(rasterState.depthFunc),
+        .stencilFront = {
+            .compare = toWebGPU(stencilState.front.stencilFunc),
+            .failOp = toWebGPU(stencilState.front.stencilOpStencilFail),
+            .depthFailOp = toWebGPU(stencilState.front.stencilOpDepthFail),
+            .passOp = toWebGPU(stencilState.front.stencilOpDepthStencilPass),
+        },
+        .stencilBack = {
+            .compare = toWebGPU(stencilState.back.stencilFunc),
+            .failOp = toWebGPU(stencilState.back.stencilOpStencilFail),
+            .depthFailOp = toWebGPU(stencilState.back.stencilOpDepthFail),
+            .passOp = toWebGPU(stencilState.back.stencilOpDepthStencilPass),
+        },
+        .stencilReadMask = 0,
+        .stencilWriteMask = stencilState.stencilWrite ? 0xFFFFFFFF : 0,
+        .depthBias = static_cast<int32_t>(polygonOffset.constant),
+        .depthBiasSlopeScale = polygonOffset.slope,
+        .depthBiasClamp = 0.0f
+    };
+    std::stringstream pipelineLabelStream;
+    pipelineLabelStream << program.name.c_str() << " pipeline";
+    const auto pipelineLabel = pipelineLabelStream.str();
+    wgpu::RenderPipelineDescriptor pipelineDescriptor{
+        .label = wgpu::StringView(pipelineLabel),
+        .layout = layout,
+        .vertex = {
+            .module = program.vertexShaderModule,
+            .entryPoint = "main",
+            .constantCount = program.constants.size(),
+            .constants = program.constants.data(),
```

Comment:
this maybe requires some investigation. For spec constants in vulkan, I had to transform c++ bool into a VK_TRUE/FALSE, which are 32-bit.